module/scheduler_components/worker.cpp

```diff
@@ -49,34 +55,143 @@ void Worker::Loop() {
     if (!msg.GetProto(req)) {
       return;
     }
+    TxnId txn_id = 0;
+    bool valid_request = true;
     switch (req.type_case()) {
       case Request::kWorker: {
-        ProcessWorkerRequest(req.worker());
+        txn_id = ProcessWorkerRequest(req.worker());
         break;
       }
       case Request::kRemoteReadResult: {
-        ProcessRemoteReadResult(req.remote_read_result());
+        txn_id = ProcessRemoteReadResult(req.remote_read_result());
         break;
       }
       default:
+        valid_request = false;
         break;
     }
+    if (valid_request) {
+      AdvanceTransaction(txn_id);
+    }
   }
   VLOG_EVERY_N(4, 5000/MODULE_POLL_TIMEOUT_MS) << "Worker " << identity_ << " is alive.";
 }
 
-void Worker::ProcessWorkerRequest(const internal::WorkerRequest& worker_request) {
+TxnId Worker::ProcessWorkerRequest(const internal::WorkerRequest& worker_request) {
   auto txn_holder = reinterpret_cast<TransactionHolder*>(worker_request.txn_holder_ptr());
   auto txn = txn_holder->GetTransaction();
-  auto txn_internal = txn->mutable_internal();
+  auto txn_id = txn->internal().id();
+  auto local_partition = config_->GetLocalPartition();
+
   RecordTxnEvent(
       config_,
-      txn_internal,
+      txn->mutable_internal(),
       TransactionEvent::ENTER_WORKER);
 
-  auto txn_id = txn_internal->id();
+  // Remove keys in remote partitions
+  auto itr = txn->mutable_read_set()->begin();
+  while (itr != txn->mutable_read_set()->end()) {
+    const auto& key = itr->first;
+    auto partition = config_->GetPartitionOfKey(key);
+    if (partition != local_partition) {
+      itr = txn->mutable_read_set()->erase(itr);
+    } else {
+      itr++;
+    }
+  }
+  itr = txn->mutable_write_set()->begin();
+  while (itr != txn->mutable_write_set()->end()) {
+    const auto& key = itr->first;
+    auto partition = config_->GetPartitionOfKey(key);
+    if (partition != local_partition) {
+      itr = txn->mutable_write_set()->erase(itr);
+    } else {
+      itr++;
+    }
+  }
+
+  // Create a state for the new transaction
+  auto state = txn_states_.emplace(
+      std::piecewise_construct,
+      std::forward_as_tuple(txn_id),
+      std::forward_as_tuple(txn_holder));
+
+  CHECK(state.second) << "Transaction " << txn_id << " has already been dispatched to this worker";
```

Comment:
This checks if Transaction state has been initialized?

Reply:
Yes. This is just another cautionary check. Later I'll replace all such checks with `DCHECK`, which would be optimized out when we compile it in release mode.

Reply:
Gotcha, didn't realize CHECK wasn't.